### PR TITLE
Enable incremental parsing via the LSP protocol

### PR DESCRIPTION
**Pull request by Ma27**
With incremental parsing, the editor only sends a "diff" of what has
changed rather than the full file. For instance, it's pretty pointless
to send all of `all-packages.nix` from `nixpkgs` for a single-letter
change.

Given the benchmark of `rnix-parser`[1] it doesn't seem as the parsing
itself is the bottleneck a few people have reported[2], so this change
only replaces affected lines and re-parses everything using
`rnix-parser`.

While the current change seems usable on local tests, it has (at
least) two known issues:

* As soon as you type `inherit`, the LSP stalls and it seems it's due to
  `rnix-parser` and steadily allocates more memory according to
  `strace(1)`. I confirmed that it's indeed a problem of `rnix-parser`
  by creating a file with `{ inherit` as content and the LSP refuses to
  respond (with incremental parsing disabled).

* Should be fixed with my second commit:
  <del>Given an expression like this:

  ``` nix
  let a = 1; in

  {
    b = a;
  }
  ```

  <del>When deleting line one and then line 2, the state ends up as `{{ }`.
  This is because the range passed to LSP looks like this:

  ```
  Range {
    start: Position {
      line: 1,
      character: 0
    },
    end: Position {
      line: 2,
      character: 0
    }
  }
  ```

  <del>This also kills the second line of the expression above and as soon as
  you delete this as well, the state will end up with bogus results.

  <del>It's a bit weird though since I got an equal message when deleting
  more lines, so I'm not entirely sure if that's just a bug in the LSP
  implementation of my neoVIM (and also the reason why I didn't just
  insert a `\n` at the right vec entries).</del>

[1] https://github.com/nix-community/rnix-parser/blob/v0.9.0/benches/all-packages.rs
[2] https://github.com/nix-community/rnix-lsp/pull/27#issuecomment-809441867

------

cc @jD91mZM2 
It may make sense if you could test this as well and see if you find more issues :) 
I'm not sure what to do about the second known issue I described btw.

Also, I guess I'll need to start writing a test suite for this, testing all of these cases when changing something isn't very nice :)

### CODE REVIEW

**Comment by Ma27** [comment]
Ok, using this for two days now and it seems as pasting seems still error-prone. Gotta investigate this a bit further. Nevertheless, feedback / testing from others is always welcome :)

File: src/main.rs [hunk 1]
```diff
@@ -232,10 +233,73 @@ impl App {
...
+                                    first.chars()
+                                        .take(start.character as usize)
+                                        .collect::<String>()
+                                );
+
+                                let suffix = String::from(
+                                    last.chars().rev()
+                                        .take(last.len() - (end.character as usize))
+                                        .collect::<String>()
+                                        .chars().rev().collect::<String>()
```
[the remark this answers is not in the document]
Comment (aaronjanse) [reply src/main.rs:276]: Perhaps `last[end.character..last.len()].to_string()`? Or maybe I'm misreading

File: src/utils.rs [hunk 2]
```diff
@@ -322,3 +322,10 @@ pub fn selection_ranges(root: &SyntaxNode, content: &str, pos: Position) -> Opti
 
     root.map(|b| *b)
 }
+
+pub fn split_vec_at<T>(v: Vec<T>, at: u64) -> (Vec<T>, Vec<T>) where T: Copy {
+    let n1 = v.clone().into_iter().take(at as usize).collect::<Vec<_>>();
+    let n2 = v.clone().into_iter().rev().take(v.len() - (at as usize)).rev().collect::<Vec<_>>();
```
Comment (aaronjanse) [inline src/utils.rs:328]: I think you could use [`.skip`](https://doc.rust-lang.org/std/iter/struct.Skip.html) or a slice here

**Comment by Ma27** [comment]
Just pushed a new implementation that's hopefully better. Will use this for a while and we'll see :)

**Comment by Ma27** [comment]
cc @jD91mZM2

**Review by aaronjanse — COMMENTED** [review]
This is great! There are a few `.get(i as usize).unwrap()` calls that could be replaced with `[i as usize]`, but I think we might actually want to keep them and eventually replace `.unwrap()` with something else so we can have a panic-free codebase.

I haven't yet had a chance to try this PR locally, but hopefully it'll help with performance on large files.

**Comment by aaronjanse** [comment]
I've tried it now, and it works great. Thank you @Ma27!

**Review by aaronjanse — APPROVED** [review]

**Comment by aaronjanse** [comment]
Feel free to merge this PR, and thank you for implementing this!

**Pull request merged**